### ISSUE TITLE
tracking issue: auto closing and selection wrapping UX shortcomings

### ISSUE THREAD

**Issue by dlqqq** [jupyterlab/jupyterlab#12700]
## Description

This issue serves to track and disambiguate the current deficiencies with the
auto-closing feature of JL.

## Terminology

I've created a list of terms to help us discuss details of the auto-closing
feature on a more granular level. Feel free to suggest changes to the
nomenclature convention I've created for this issue.

- enclosure: a range enclosed by two characters which form some syntactically
valid body. e.g. `""`, `()`, `[]`, `{}`.

- nested enclosure: an enclosure within another enclosure. E.g. in the
  expression `([])`, `[]` is a nested enclosure.

- opening character: the first character of an enclosure, e.g. `(`, `[`, `{`.

- closing character: the last character of an enclosure, e.g. `)`, `]`, `}`.

- auto-closing (AC): feature where when a user inputs an opening character, the
  IDE intelligently inserts closing characters when applicable. Collectively
  refers to the feature enabled by `autoClosingQuotes` and `autoClosingBrackets`.

- selection wrapping (SW): feature where a user can select a range of text,
  input an opening character, and create a new enclosure around the selected range
  with the cursor moved after the end of the range. E.g. selecting the range
  `muffin`, inputting `"` results in `"muffin"` and the cursor moved to the end of
  that range.

## 1. Deleting a nested enclosure

From thread in: https://github.com/jupyterlab/jupyterlab/issues/5741

When attempting to delete a nested enclosure via backspacing, AC erroneously
deletes an extra closing character, resulting in an invalid expression.

https://user-images.githubusercontent.com/44106031/173690140-b27490e9-1352-4c01-809b-679e24e8d14f.mov

Expected behavior: the expression should be `f()` and not `f(`.

What's happening here is that when the cursor is preceded by an opening
character and succeeded by a matching opening character, the enclosure is always
deleted even if its deletion forms a syntactically invalid expression.

## 2. Inputting closing character with cursor preceding closing character

From thread in: https://github.com/jupyterlab/jupyterlab/issues/5741

When inputting a closing character, AC will always ignore user input if the
cursor precedes a closing character. This UX deficiency is most noticable when
wrapping a function argument within another function.

https://user-images.githubusercontent.com/44106031/173692284-8ed1a861-0ea6-42b8-ae6b-ac568eafbdf7.mov

Expeected behavior: user should only have to input `)` once instead of 3 times

Interestingly, VSCode similarly ignores closing characters when the cursor is
succeeded by closing characters, but always respects user input if the user did
not type the original closing characters in the same typing "stream". See below.
When I deselect and reselect the line in the second example, VSCode respects my
input.

https://user-images.githubusercontent.com/44106031/173693849-9100d772-50f0-4139-9ffa-043102c98f0c.mov

I believe this is the behavior we should aim to emulate.

## 3. Selection wrapping should be separate from auto-closing

Tracking issue: https://github.com/jupyterlab/jupyterlab/issues/9897

Users should be able to enjoy the benefits of selection wrapping without
enabling AC. Looks like this is an upstream issue with the way CM5 handles
configuration.

## 4. Python F-Strings and successive quotes

Tracking issue: https://github.com/codemirror/codemirror5/issues/2344

In summary, the current AC implementation within CM5 does not correctly handle
Python F-strings and successive quotes.

- When typing `f"`, AC fails to add recognize `"` as an opening character and does not insert a closing `"`.

- When typing `'''`, AC adds too many closing characters. It is currently not
  possible to type exactly three single quotes (e.g. for heredoc comments).

https://user-images.githubusercontent.com/44106031/173694727-d679c01d-d04f-41f1-be19-24d56477c6e4.mov

## Context

- Operating System and version: all
- Browser and version: all
- JupyterLab version: 3.0+

## References

- [VSCode blog on bracket coloring](https://code.visualstudio.com/blogs/2021/09/29/bracket-pair-colorization)

**Comment by jweill-aws** [comment]
Triage notes: Discussed at the 6/15 JupyterLab meeting. Need to check the behavior in CodeMirror 6, to see whether it handles any of these problems.

**Comment by psychemedia** [comment]
What's the status of this In JupyterLab 4; whatever user settings I tweak, I can't get a behaviour other than deleting selected text and replacing it with a single quote or bracket?
